File: haskell/grep_std.hs
-- haskgrep_std - idiomatic single-threaded Haskell.
-- Strict Data.ByteString (Word8 octets) all the way: byte-exact offsets,
-- breakSubstring for the literal search, System.Directory for the walk.
-- No threads, no hand-rolled SIMD/syscalls. stdlib only.
module Main (main) where

import qualified Data.ByteString as BS
import           Data.ByteString (ByteString)
import           Data.Word (Word8)
import           Data.IORef
import           System.Environment (getArgs)
import           System.Exit (exitWith, ExitCode (..))
import           System.IO
import           System.Directory (doesDirectoryExist, listDirectory, pathIsSymbolicLink)
import           System.FilePath ((</>))
import qualified Data.ByteString.Char8 as BC

-- Config carried explicitly (single-threaded, but keep it pure-ish).
data Cfg = Cfg
  { cfgPat    :: !ByteString   -- needle as given
  , cfgLPat   :: !ByteString   -- ASCII-lowercased needle (for -i)
  , cfgCI     :: !Bool
  , cfgMulti  :: !Bool
  , cfgMatch  :: !(IORef Bool)
  }

-- ASCII-only, length-preserving lowercase. Matches grep -iF; NOT Unicode
-- Data.Char.toLower (which could change byte length / offsets).
asciiLower :: ByteString -> ByteString
asciiLower = BS.map (\b -> if b >= 65 && b <= 90 then b + 32 else b)

nul :: Word8
nul = 0

nl :: Word8
nl = 10

searchFile :: Cfg -> FilePath -> IO ()
searchFile cfg path = do
  dat <- BS.readFile path
  let len  = BS.length dat
      peek = min len 65536
  if BS.elem nul (BS.take peek dat)
    then return ()                       -- binary, skip
    else do
      let (hay, needle)
            | cfgCI cfg = (asciiLower dat, cfgLPat cfg)
            | otherwise = (dat, cfgPat cfg)
          loop pos
            | pos >= len = return ()      -- pos < len guard: empty-pattern fix
            | otherwise =
                let (pre, post) = BS.breakSubstring needle (BS.drop pos hay)
                in if BS.null post && not (BS.null needle)
                     then return ()       -- not found
                     else if BS.null post
                            then return () -- empty needle, no occurrence past pos
                            else do
                              let m  = pos + BS.length pre
                                  ls = lastNlBefore dat m + 1
                                  le = firstNlAt dat m len
                              writeIORef (cfgMatch cfg) True
                              if cfgMulti cfg
                                then BS.hPutStr stdout (BC.pack path) >> BS.hPutStr stdout (BS.singleton 58)
                                else return ()
                              BS.hPutStr stdout (BS.take (le - ls) (BS.drop ls dat))
                              BS.hPutStr stdout (BS.singleton nl)
                              loop (le + 1)
      loop 0

-- last index of '\n' strictly before m, or -1.
lastNlBefore :: ByteString -> Int -> Int
lastNlBefore dat m =
  case BS.elemIndexEnd nl (BS.take m dat) of
    Just i  -> i
    Nothing -> -1

-- first index of '\n' at/after m, or `len`.
firstNlAt :: ByteString -> Int -> Int -> Int
firstNlAt dat m len =
  case BS.elemIndex nl (BS.drop m dat) of
    Just j  -> m + j
    Nothing -> len

-- Recursive walk: regular files only, never follow symlinks.
walk :: Cfg -> FilePath -> IO ()
walk cfg p = do
  isLink <- pathIsSymbolicLink p
  if isLink
    then return ()
    else do
      isDir <- doesDirectoryExist p
      if isDir
        then do
          entries <- listDirectory p
          mapM_ (\e -> walk cfg (p </> e)) entries
        else searchFile cfg p

usage :: IO a
usage = do
  hPutStr stderr "usage: haskgrep [-r] [-i] PATTERN PATH...\n"
  exitWith (ExitFailure 2)

-- Parse args: flags may combine (-ri); -- ends options; first non-flag = PATTERN.
parseArgs :: [String] -> Maybe (Bool, Bool, ByteString, [FilePath])
parseArgs = go False False Nothing [] False
  where
    go ci rec mpat paths _ [] =
      case mpat of
        Just pat | not (null paths) -> Just (ci, rec, pat, reverse paths)
        _                           -> Nothing
    go ci rec mpat paths noMore (a:as)
      | not noMore && a == "--" = go ci rec mpat paths True as
      | not noMore && length a >= 2 && head a == '-' =
          case parseFlags ci rec (tail a) of
            Just (ci', rec') -> go ci' rec' mpat paths noMore as
            Nothing          -> Nothing      -- unknown flag
      | Nothing <- mpat = go ci rec (Just (BC.pack a)) paths noMore as
      | otherwise       = go ci rec mpat (a : paths) noMore as
    parseFlags ci rec [] = Just (ci, rec)
    parseFlags ci rec (c:cs) = case c of
      'i' -> parseFlags True rec cs
      'r' -> parseFlags ci True cs
      _   -> Nothing

main :: IO ()
main = do
  args <- getArgs
  case parseArgs args of
    Nothing -> usage
    Just (ci, recursive, pat, paths) -> do
      hSetBinaryMode stdout True
      hSetBuffering stdout (BlockBuffering (Just 65536))
      matchRef <- newIORef False
      let multi = recursive || length paths > 1
          cfg = Cfg { cfgPat = pat
                    , cfgLPat = asciiLower pat
                    , cfgCI = ci
                    , cfgMulti = multi
                    , cfgMatch = matchRef
                    }
      mapM_ (processPath cfg recursive) paths
      hFlush stdout
      matched <- readIORef matchRef
      exitWith (if matched then ExitSuccess else ExitFailure 1)

processPath :: Cfg -> Bool -> FilePath -> IO ()
processPath cfg recursive p = do
  isDir <- doesDirectoryExist p
  if isDir
    then if recursive then walk cfg p else return ()   -- dir w/o -r: skip
    else searchFile cfg p

File: elixir/grep_std.exs
# exgrep_std - idiomatic single-threaded Elixir on the BEAM.
#
# The exotic VM in the set. Immutable binaries all the way (byte-exact offsets),
# :binary.match/2 (a C BIF) for the literal search -- NOT Regex (which would treat
# the pattern as a regular expression and isn't the fast literal scan we want).
# File.ls!/File.lstat for the walk; symlinks are never followed. No threads.
#
# Whole-file read via File.read!. NUL-in-first-64KB binary skip. ASCII-only
# case folding for -i (NOT String.downcase, which is Unicode and could change
# byte length / offsets -- we map bytes A-Z -> a-z ourselves to match grep -iF).

defmodule ExGrep do
  # ASCII-only, length-preserving lowercase: byte map A-Z (65..90) -> a-z.
  # Done with :binary.bin_to_list + map; immutable, builds a fresh binary.
  def ascii_lower(bin) do
    bin
    |> :binary.bin_to_list()
    |> Enum.map(fn b -> if b >= ?A and b <= ?Z, do: b + 32, else: b end)
    |> :binary.list_to_bin()
  end

  # last index of \n strictly before m, or -1  => line start = that + 1
  defp line_start(data, m) do
    case :binary.matches(:binary.part(data, 0, m), "\n") do
      [] -> 0
      list -> (List.last(list) |> elem(0)) + 1
    end
  end

  # first index of \n at/after m, or len
  defp line_end(data, m, len) do
    case :binary.match(data, "\n", scope: {m, len - m}) do
      :nomatch -> len
      {pos, _} -> pos
    end
  end

  def search_file(cfg, path) do
    case File.read(path) do
      {:error, _} ->
        false

      {:ok, data} ->
        len = byte_size(data)
        peek = min(len, 65536)

        binary? =
          peek > 0 and :binary.match(data, <<0>>, scope: {0, peek}) != :nomatch

        cond do
          binary? ->
            false

          len == 0 ->
            false

          # Empty pattern matches every line (grep -F "" behaviour).
          # :binary.match rejects an empty needle, so emit each line directly.
          cfg.pat == "" ->
            emit_all_lines(cfg, path, data, len, 0)
            true

          true ->
            {hay, needle} =
              if cfg.ci, do: {ascii_lower(data), cfg.lpat}, else: {data, cfg.pat}

            scan(cfg, path, data, hay, needle, len, 0, false)
        end
    end
  end

  defp emit_all_lines(_cfg, _path, _data, len, pos) when pos >= len, do: :ok

  defp emit_all_lines(cfg, path, data, len, pos) do
    le = line_end(data, pos, len)
    line = :binary.part(data, pos, le - pos)

    if cfg.multi do
      IO.binwrite(:stdio, [path, ?:, line, ?\n])
    else
      IO.binwrite(:stdio, [line, ?\n])
    end

    emit_all_lines(cfg, path, data, len, le + 1)
  end

  # Drive :binary.match across the haystack, printing each matching line once
  # then resuming past its end (le + 1). Returns whether anything matched.
  defp scan(cfg, path, data, hay, needle, len, pos, matched) do
    if pos >= len do
      matched
    else
      case :binary.match(hay, needle, scope: {pos, len - pos}) do
        :nomatch ->
          matched

        {m, _mlen} ->
          ls = line_start(data, m)
          le = line_end(data, m, len)
          line = :binary.part(data, ls, le - ls)

          if cfg.multi do
            IO.binwrite(:stdio, [path, ?:, line, ?\n])
          else
            IO.binwrite(:stdio, [line, ?\n])
          end

          scan(cfg, path, data, hay, needle, len, le + 1, true)
      end
    end
  end

  # Recursive walk: regular files only, never follow symlinks.
  def walk(cfg, path) do
    case File.lstat(path) do
      {:ok, %File.Stat{type: :symlink}} ->
        false

      {:ok, %File.Stat{type: :directory}} ->
        case File.ls(path) do
          {:ok, entries} ->
            Enum.reduce(entries, false, fn e, acc ->
              walk(cfg, Path.join(path, e)) or acc
            end)

          {:error, _} ->
            false
        end

      {:ok, %File.Stat{type: :regular}} ->
        search_file(cfg, path)

      _ ->
        false
    end
  end

  def process_path(cfg, recursive, path) do
    case File.stat(path) do
      {:ok, %File.Stat{type: :directory}} ->
        if recursive, do: walk(cfg, path), else: false

      {:ok, _} ->
        search_file(cfg, path)

      {:error, _} ->
        false
    end
  end

  # Parse args: flags may combine (-ri); -- ends options; first non-flag = PATTERN.
  defp strip_launcher_dashes(["--" | rest]), do: rest
  defp strip_launcher_dashes(args), do: args

  def parse_args(args), do: parse(args, false, false, nil, [], false)

  defp parse([], ci, rec, pat, paths, _no_more) do
    case {pat, paths} do
      {nil, _} -> :error
      {_, []} -> :error
      {p, ps} -> {:ok, ci, rec, p, Enum.reverse(ps)}
    end
  end

  defp parse(["--" | rest], ci, rec, pat, paths, false),
    do: parse(rest, ci, rec, pat, paths, true)

  defp parse([a | rest], ci, rec, pat, paths, false = nm)
       when binary_part(a, 0, 1) == "-" and byte_size(a) >= 2 do
    case parse_flags(String.to_charlist(binary_part(a, 1, byte_size(a) - 1)), ci, rec) do
      {:ok, ci2, rec2} -> parse(rest, ci2, rec2, pat, paths, nm)
      :error -> :error
    end
  end

  defp parse([a | rest], ci, rec, nil, paths, nm),
    do: parse(rest, ci, rec, a, paths, nm)

  defp parse([a | rest], ci, rec, pat, paths, nm),
    do: parse(rest, ci, rec, pat, [a | paths], nm)

  defp parse_flags([], ci, rec), do: {:ok, ci, rec}
  defp parse_flags([?i | cs], _ci, rec), do: parse_flags(cs, true, rec)
  defp parse_flags([?r | cs], ci, _rec), do: parse_flags(cs, ci, true)
  defp parse_flags(_, _ci, _rec), do: :error

  def main(args) do
    # The bin/ launcher execs `elixir grep_std.exs -- "$@"`; that leading "--"
    # shields elixir's own arg parser from user flags and DOES appear in argv.
    # Strip exactly one so a user-supplied "--" is still honored by parse_args.
    args = strip_launcher_dashes(args)

    case parse_args(args) do
      :error ->
        IO.binwrite(:stderr, "usage: exgrep [-r] [-i] PATTERN PATH...\n")
        System.halt(2)

      {:ok, ci, recursive, pat, paths} ->
        multi = recursive or length(paths) > 1

        cfg = %{
          pat: pat,
          lpat: ascii_lower(pat),
          ci: ci,
          multi: multi
        }

        matched =
          Enum.reduce(paths, false, fn p, acc ->
            process_path(cfg, recursive, p) or acc
          end)

        System.halt(if matched, do: 0, else: 1)
    end
  end
end

ExGrep.main(System.argv())
